doc/source/ray-core/ray-generator.rst

```diff
@@ -0,0 +1,210 @@
+
+
+The above Ray generator yields the output every 5 seconds 5 times.
+When you use a normal Ray task, you have to wait 25 seconds to access the output. 
+With a Ray generator, the caller can access the object refeference
+before the task ``f`` finishes.
+
+**Ray generator is useful when**
+
+- You want to reduce heap memory or object store memory usage by yielding and GC the output before task finishes.
+- You are familiar with Python generator and want the equivalent programming models.
+
+**Ray generator is used by Ray libraries to support stremaing use cases**
+
+- :ref:`Ray serve <rayserve>` uses Ray generators to support :ref:`streaming responses <serve-http-streaming-response>`.
+- :ref:`Ray data <data>` is a streaming data processing library, and it uses Ray generators to control and reduce concurrent memory usages.
+
+**Ray generator works with existing Ray APIs seamlessly**
+
+- Ray generators can be used in both actor and non-actor tasks.
+- Ray generators work with all actor execution model, including :ref:`threaded actors <threaded-actors>` and :ref:`async actors <async-actors>`.
+- Ray generators work with built-in :ref:`fault tolerance features <fault-tolerance>` such as retry or lineage reconstruction.
+- Ray generators work with Ray APIs such as :ref:`ray.wait <generators-wait>`, :ref:`ray.cancel <generators-cancel>`, etc.
+
+Getting Started
+---------------
+You have to define a Python generator function and specify ``num_returns="streaming"`` 
+to create a Ray generators.
+
+.. literalinclude:: doc_code/streaming_generator.py
+    :language: python
+    :start-after: __streaming_generator_define_start__
+    :end-before: __streaming_generator_define_end__
+
+The ray generator task returns ``StreamingObjectRefGenerator`` (API is subject to change), which is
+compatible to generator and async generator APIs. It means 
+``next``, ``__iter__``, ``__anext__``, ``__aiter__`` APIs are accessible from the class. 
+
+Whenever a task invokes ``yield``, a corresponding output is ready and availabale from a generator as a Ray object reference. 
+You can call ``next(gen)`` to obtain a object reference.
+If ``next`` has no more item to generate, it raises ``StopIteration``. If ``__anext__`` has no more item to generate, it raises
+``StopAsyncIteration``
+
+``next`` API will block its thread until a next object reference is generated from the task via ``yield``.
+Since the ``StreamingObjectRefGenerator`` is just a Python generator, you can also simply use a for loop to
+iterate object references. 
+
+If you want to avoid blocking a thread, you can either use asyncio or :ref:`ray.wait API <generators-wait>`.
+
+.. literalinclude:: doc_code/streaming_generator.py
+    :language: python
+    :start-after: __streaming_generator_execute_start__
+    :end-before: __streaming_generator_execute_end__
+
+.. note::
+
+    For a normal Python generator, a generator function is paused and resumed when ``next`` function is
```

Comment:
```suggestion
    Python pauses a generator function is paused and resumes it with ``next`` function
```

Reply:
Not sure if this rewrite is correct any more, but the suggestion is to use active voice. Be clear about what/who is doing the pausing and resuming.